Add example for HazelcastDataConnection (#742)
hazelcast/hz-docs: docs/modules/data-connections/pages/data-connections-configuration.adoc

--
====

[[Hazelcast]]
=== Example Hazelcast Data Connection

This example configuration shows a data connection to a remote Hazelcast cluster.
You can use a Hazelcast data connection from the Pipeline API in link:https://docs.hazelcast.org/docs/{full-version}/javadoc/com/hazelcast/jet/pipeline/Sources.html#remoteMapJournal-java.lang.String-com.hazelcast.jet.pipeline.DataConnectionRef-com.hazelcast.jet.pipeline.JournalInitialPosition-com.hazelcast.function.FunctionEx-com.hazelcast.function.PredicateEx-[Sources#remoteMapJournal] source.

NOTE: Currently, no SQL connector is available for Hazelcast data connections. This means that although you can xref:sql:create-data-connection.adoc[create a data connection in SQL], you cannot yet use it in SQL, for example, in a mapping statement. 

> is no SQL connector for remote Hazelcast available.

Reviewer: ```suggestion
```
I think this text just got left behind after the note was reworded.

[tabs]
====
XML::
+
--
[source,xml]
----
<hazelcast>
  <data-connection name="my-remote-hazelcast">
    <type>Hz</type>
    <properties>
      <property name="client_xml"> <1>
        <![CDATA[ <2>
          <?xml version="1.0" encoding="UTF-8"?>
          <hazelcast-client xmlns="http://www.hazelcast.com/schema/client-config"
              xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
              xsi:schemaLocation="http://www.hazelcast.com/schema/client-config

>               http://www.hazelcast.com/schema/client-config/hazelcast-client-config-5.3.xsd">

Author: @Serdaro @angelasimms 
This should match the minor version of running hazelcalcast cluster, we have `{full-version}` but I didn't find anything like `{minor-version}`

            <cluster-name>dev</cluster-name>
              <network>
                <cluster-members>
                  <address>172.17.0.2:5701</address>
                </cluster-members>
              </network>
          </hazelcast-client>
        ]]>
      </property>
    </properties>
    <shared>true</shared>
  </data-connection>
</hazelcast>
----
<1> (Required) Exactly one of `client_xml`, `client_yml`, `client_xml_path`, `client_yml_path` must be specified.
<2> Hazelcast client configuration to connect to a remote cluster. See xref:clients:java.adoc#configuring-java-client[Configuring Java Client].
External file can be specified using `client_xml_path` property instead of embedded configuration file.
--

YAML::
+
--
[source,yaml]
----
hazelcast:
  data-connection:
    my-remote-hazelcast:
      type: Hz
      properties:
        client_yml: | <1>
          hazelcast-client: <2>
            cluster-name: dev
            network:
              cluster-members:
                - 172.17.0.2:5701
      shared: true
----

> <1> (Required) Exactly one of `client_xml`, `client_yml`, `client_xml_path`, `client_yml_path` must be specified.

Reviewer: ```suggestion
<1> (Required) Specify exactly one of `client_xml`, `client_yml`, `client_xml_path`, `client_yml_path`.
```

> <2> Hazelcast client configuration to connect to a remote cluster. See xref:clients:java.adoc#configuring-java-client[Configuring Java Client]].

Reviewer: ```suggestion
<2> Hazelcast client configuration to connect to a remote cluster. See xref:clients:java.adoc#configuring-java-client[Configuring Java Client].
```

> External file can be specified using `client_yml_path` property instead of embedded configuration file.

Reviewer: ```suggestion
You can specify an external file with the `client_yml_path` property instead of using an embedded configuration file.
```

--

Java::
+
--
[source,java]
----
config
  .addDataConnectionConfig(
    new DataConnectionConfig()
      .setName("my-remote-hazelcast")
      .setType("Hz")
      .setProperty(
              "client_xml", <1>
              "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" + <2>
              "<hazelcast-client xmlns=\"http://www.hazelcast.com/schema/client-config\"" +
              "      xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"" +
              "      xsi:schemaLocation=\"http://www.hazelcast.com/schema/client-config" +
              "      http://www.hazelcast.com/schema/client-config/hazelcast-client-config-5.3.xsd\">" +
              "    " +
              "    <cluster-name>dev</cluster-name>" +
              "    <network>" +
              "        <cluster-members>" +
              "            <address>172.17.0.2:5701</address>" +
              "        </cluster-members>" +
              "    </network>" +
              "</hazelcast-client>")
      .setShared(true)
  );
----

> <1> (Required) Exactly one of `client_xml`, `client_yml`, `client_xml_path`, `client_yml_path` must be specified.

Reviewer: ```suggestion
<1> (Required) Specify exactly one of `client_xml`, `client_yml`, `client_xml_path`, `client_yml_path`.
```

<2> Hazelcast client configuration to connect to a remote cluster. See xref:clients:java.adoc#configuring-java-client[Configuring Java Client].

> External file can be specified using `client_xml_path` or `client_yml_path` property instead of embedded configuration file.

Reviewer: ```suggestion
You can specify an external file with the `client_xml_path` or `client_yml_path` property instead of using an embedded configuration file.
```

--
SQL::
+
--
Data connections created in SQL behave differently to those defined in members' configuration files or in Java.

> - To retain SQL-defined data connections after a cluster restart, you must enable SQL catalog persistence. This feature is available in the Enterprise Edition.

Reviewer: ```suggestion
- To retain SQL-defined data connections after a cluster restart, you must enable xref:storage:configuring-persistence.adoc#sql[SQL metadata persistence]. This feature is available in the Enterprise Edition.
```
I added more info about SQL catalog persistence yesterday. Other data connection 'notes' were updated as part of this work.

- You can create or drop a data connection using SQL commands. To update a data connection, you need to drop and then recreate it.

> NOTE: While Hazelcast data connection can be created using SQL command it cannot yet be used from SQL, because there

Reviewer: I don't think we need this note as the same text already appears the intro text before the tabbed examples. Here's a preview of the page https://deploy-preview-742--hardcore-allen-f5257d.netlify.app/hazelcast/5.4-snapshot/data-connections/data-connections-configuration#Hazelcast

is no SQL connector for remote Hazelcast available.

[source,SQL]
----
CREATE DATA CONNECTION "my-hazelcast-cluster"
TYPE Hz
SHARED
OPTIONS (
    'client_xml'= <1>
    '<?xml version="1.0" encoding="UTF-8"?> <2>
     <hazelcast-client xmlns="http://www.hazelcast.com/schema/client-config"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://www.hazelcast.com/schema/client-config
         http://www.hazelcast.com/schema/client-config/hazelcast-client-config-5.3.xsd">

       <cluster-name>dev</cluster-name>
       <network>
         <cluster-members>
	   <address>172.17.0.2:5701</address>
         </cluster-members>
       </network>
      </hazelcast-client>'
);
----

> <1> (Required) Exactly one of `client_xml`, `client_yml`, `client_xml_path`, `client_yml_path` must be specified.

Reviewer: ```suggestion
<1> (Required) Specify exactly one of `client_xml`, `client_yml`, `client_xml_path`, `client_yml_path`.
```

<2> Hazelcast client configuration to connect to a remote cluster. See xref:clients:java.adoc#configuring-java-client[Configuring Java Client].

> External file can be specified using `client_xml_path` or `client_yml_path` property instead of embedded configuration file.

Reviewer: ```suggestion
You can specify an external file with the `client_xml_path` or `client_yml_path` property instead of using an embedded configuration file.
```

--

> ====

Reviewer: Link to the HazelcastDataConnection java docs is missing from the `shared` configuration option.
